add traffic to CostMatrix

nilsnolde opened the pull request:
fixes https://github.com/valhalla/valhalla/issues/4038

Extending the work done in https://github.com/valhalla/valhalla/pull/3795, I added time awareness to the CostMatrix algo as well.

A few high-level things to mention:
- currently we **only** support adding `depart_at`/`current` time spec, i.e. only on origin locations. I left a TODO for `arrive_by` which could be done (also for bidir A*), but not a priority for me at least and also quite a bit more work/complexity
- the whole decision model in `thor/matrix_action` changed a bit. It still preserves the previous way we decided on the matrix algo, so nothing will change for Valhalla deployments which don't have any time info
- unless `prioritize_bidirectional: true` is set (and of course plus a valid time on the sources/targets), we'll favor the more accurate (albeit much slower) TimeDistance matrix algo
- to trigger a CostMatrix with time awareness:
  - add `"prioritize_bidirectional": true` in the root of the request
  - add at least one `date_time` string to an origin location or set the global `date_time.type = 1/0`

nilsnolde commented:
I'll leave it here for now. Next week I'll be on vacation, I'll comment more after. Feel free to have a glance of course:)

nilsnolde commented on src/sif/recost.cc:
```diff
@@ -98,9 +98,9 @@ void recost_forward(baldr::GraphReader& reader,
-        (!costing.Allowed(edge, !next_id.Is_Valid(), label, tile, edge_id, localtime,
-                          offset_time.timezone_index, time_restrictions_TODO) &&
-         !ignore_access)) {
+        (!ignore_access &&
+         !costing.Allowed(edge, !next_id.Is_Valid(), label, tile, edge_id, localtime,
+                          offset_time.timezone_index, time_restrictions_TODO))) {
```
tiny nit: just changed around the order of calls, ignore_accees is much cheaper

nilsnolde commented on src/thor/bidirectional_astar.cc:
```diff
@@ -1276,6 +1276,9 @@ std::vector<std::vector<PathInfo>> BidirectionalAStar::FormPath(GraphReader& gra
     }
 
     // recost edges in final path; ignore access restrictions
+    // TODO: actually we should not ignore access restrictions: if the reverse path
+    //   traversed a closed edge due to time restrictions, we could do a mini traversal
+    //   to circumvent the closed edge(s)
```
it's not the nicest patch, but still nicer than navigating over an edge we know is closed

nilsnolde commented on src/thor/costmatrix.cc:
```diff
@@ -108,12 +117,14 @@ void CostMatrix::clear() {
-    const float max_matrix_distance) {
+    const sif::travel_mode_t mode,
+    const float max_matrix_distance,
+    const bool has_time,
```
we tell it when there's a valid time. we compute it in `matrix_action` where we already check if any time is valid

nilsnolde commented on src/thor/costmatrix.cc:
```diff
@@ -890,5 +930,111 @@ void CostMatrix::SetTargets(baldr::GraphReader& graphreader,
+    };
+
+    Cost new_cost{0.f, 0.f};
+    const auto label_cb = [&new_cost](const EdgeLabel& label) { new_cost = label.cost(); };
```
we only need to update the cost when recosting the path, not path necessary. once we want to tackle https://github.com/valhalla/valhalla/issues/3920, we can do it similar to bidir a*

nilsnolde commented on src/thor/costmatrix.cc:
```diff
@@ -890,5 +930,111 @@ void CostMatrix::SetTargets(baldr::GraphReader& graphreader,
+    }
+
+    auto edge_itr = path_edges.begin();
+    const auto edge_cb = [&edge_itr, &path_edges]() {
+      return (edge_itr == path_edges.end()) ? GraphId{} : (*edge_itr++);
+    };
```
same as bidir a*, only the label callback changes

nilsnolde commented on src/thor/matrix_action.cc:
```diff
@@ -14,39 +14,6 @@ using namespace valhalla::baldr;
-
-// return true if any location had a time set
-// also disable time if it doesn't make sense computationally
-bool has_time(Api& request) {
```
changed it to `check_matrix_time` and moved it to `matrix_common.h`

nilsnolde commented on src/thor/matrix_action.cc:
```diff
@@ -93,33 +63,40 @@ std::string thor_worker_t::matrix(Api& request) {
+  bool has_time =
+      check_matrix_time(request,
+                        options.prioritize_bidirectional() ? MatrixType::Cost : MatrixType::TimeDist);
+  if (has_time && !options.prioritize_bidirectional() && !force_costmatrix) {
```
do a timedistance matrix, no matter the costing model, if:
- there's time
- `prioritize_bidirectional` is false
- the server doesn't have `COST_MATRIX` defined

nilsnolde commented on src/thor/matrix_action.cc:
```diff
@@ -93,33 +63,40 @@ std::string thor_worker_t::matrix(Api& request) {
+                        options.prioritize_bidirectional() ? MatrixType::Cost : MatrixType::TimeDist);
+  if (has_time && !options.prioritize_bidirectional() && !force_costmatrix) {
+    return tyr::serializeMatrix(request, timedistancematrix(), distance_scale, MatrixType::TimeDist);
+  } else if (has_time && options.prioritize_bidirectional()) {
```
otherwise, if there's a valid time AND `prioritize_bidirectional` is true, we do a CostMatrix

nilsnolde commented on src/thor/matrix_action.cc:
```diff
@@ -93,33 +63,40 @@ std::string thor_worker_t::matrix(Api& request) {
+    return tyr::serializeMatrix(request, timedistancematrix(), distance_scale, MatrixType::TimeDist);
+  } else if (has_time && options.prioritize_bidirectional()) {
+    return tyr::serializeMatrix(request, costmatrix(has_time), distance_scale, MatrixType::Cost);
+  } else if (matrix_type == MatrixType::Cost) {
+    return tyr::serializeMatrix(request, costmatrix(has_time), distance_scale, matrix_type);
+  } else {
+    return tyr::serializeMatrix(request, timedistancematrix(), distance_scale, matrix_type);
```
these two are the same as before

nilsnolde commented on src/thor/optimized_route_action.cc:
```diff
@@ -28,9 +28,11 @@ void thor_worker_t::optimized_route(Api& request) {
+  std::vector<TimeDistance> td =
+      costmatrix.SourceToTarget(*options.mutable_sources(), *options.mutable_targets(), *reader,
+                                mode_costing, mode, max_matrix_distance.find(costing)->second,
+                                check_matrix_time(request, MatrixType::Cost),
+                                options.date_time_type() == Options::invariant);
```
also enable time awareness for `/optimized`

nilsnolde commented on src/thor/timedistancematrix.cc:
```diff
@@ -22,37 +22,6 @@ static bool IsTrivial(const uint64_t& edgeid,
 }
-
-// Will return a destination's date_time string
-std::string GetDateTime(const std::string& origin_dt,
```
moved this to `matrix_common.h` as well so we can use it in both

nilsnolde commented on src/tyr/matrix_serializer.cc:
```diff
@@ -2,7 +2,7 @@
 #include "baldr/json.h"
 #include "proto_conversions.h"
-#include "thor/costmatrix.h"
+#include "thor/matrix_common.h"
```
I moved most common things, which were actually previously defined in `costmatrix.h` to `matrix_common.h` as the import pattern was a bit weird otherwise (timedistance.h matrix would import costmatrix.h..)

nilsnolde commented on src/tyr/matrix_serializer.cc:
```diff
@@ -71,6 +72,8 @@ json::MapPtr serialize(const Api& request,
   }
   json->emplace("durations", time);
   json->emplace("distances", distance);
+  json->emplace("algorithm",
+                std::string(matrix_type == MatrixType::Cost ? "costmatrix" : "timedistancematrix"));
```
add a `algorithm` field to the response

nilsnolde commented on valhalla/thor/costmatrix.h:
```diff
@@ -27,24 +29,6 @@ constexpr float kCostThresholdBicycleDivisor =
     56.0f; // 200 km distance threshold will result in a cost threshold of ~3600 (1 hour)
 constexpr float kCostThresholdPedestrianDivisor =
     28.0f; // 200 km distance threshold will result in a cost threshold of ~7200 (2 hours)
-constexpr float kMaxCost = 99999999.9999f;
-
-// Time and Distance structure
-struct TimeDistance {
```
moved these to `matrix_common.h` as well

nilsnolde commented on valhalla/thor/matrix_common.h:
```diff
@@ -15,15 +15,18 @@
 #include <valhalla/sif/dynamiccost.h>
 #include <valhalla/sif/edgelabel.h>
 #include <valhalla/thor/astarheuristic.h>
-#include <valhalla/thor/costmatrix.h>
```
this was strange :smile:

nilsnolde commented on src/thor/matrix_action.cc:
```diff
@@ -93,33 +63,40 @@ std::string thor_worker_t::matrix(Api& request) {
-            time_distances = costmatrix();
-          } else {
-            time_distances = timedistancematrix();
+          if (options.sources().size() < kCostMatrixThreshold ||
+              options.targets().size() < kCostMatrixThreshold) {
```
<= instead of <

kevinkreiser commented on src/thor/costmatrix.cc:
```diff
@@ -890,5 +929,111 @@ void CostMatrix::SetTargets(baldr::GraphReader& graphreader,
 }
 
+// Form the path from the adjacency list.
+void CostMatrix::RecostPaths(GraphReader& graphreader,
```
it would be cool to move the inner loop of recosting a single src and tgt to a common area like path info or somethign else that makes sense so that this and bidir a* can share it

nilsnolde replied:
left a todo

nilsnolde commented on valhalla/thor/costmatrix.h:
```diff
@@ -247,6 +241,42 @@ class CostMatrix {
+   */
+  std::vector<baldr::TimeInfo>
+  SetOriginTimes(google::protobuf::RepeatedPtrField<valhalla::Location>& origins,
+                 baldr::GraphReader& reader) {
```
make it const function

nilsnolde replied:
move to anon namespace in the src

nilsnolde replied:
ah I put it in the header bcs it needs access to `tz_cache_` member.

nilsnolde commented:
Addressed the review comments from yesterday @kevinkreiser :

- https://github.com/valhalla/valhalla/pull/4071/commits/9419e4beb8c322e3b17da4549de2455722dccf82: re-arranged the decision model for which algo to use in `matrix_action` a little
- https://github.com/valhalla/valhalla/pull/4071/commits/ada29d708ead4e1d48f24312c5a3ade654ea683d: Also added warnings if the config forces the algorithm but the request im-/explicitly suggested a different algorithm.
- https://github.com/valhalla/valhalla/pull/4071/commits/49ff04b3049d38cac669a80e4d4ca147ab29cc9a: minor things like leaving a todo & changing `<` to `<=` in one place

nilsnolde commented:
There was a minor conflict in costmatrix.cc. After CI is done, this is good to merge from my side @kevinkreiser :)

kevinkreiser commented:
looks like setsources is missing the timeinfo arg

nilsnolde commented:
Yeah there were a few code merge conflicts and I shouldn't have tried solving those in the Github UI, it made a bit of a mess. That function was called twice, old & new way, so it threw.

I fixed it.

kevinkreiser commented on src/thor/costmatrix.cc:
```diff
@@ -750,7 +787,13 @@ void CostMatrix::BackwardSearch(const uint32_t index, GraphReader& graphreader)
+                            const std::vector<baldr::TimeInfo>& time_infos) {
+  // Allocate edge labels and edge status
+  source_count_ = sources.size();
+  source_edgelabel_.resize(source_count_);
+  source_edgestatus_.resize(source_count_);
```
arent these now redundant with `Initialize`?
